src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java

```diff
@@ -1,202 +1,326 @@
-    esDao.setMimeType(getMimeType());
-    return esDao;
-  }
-
-  @Extension
-  public static class ElasticSearchDescriptor extends LogstashIndexerDescriptor
-  {
     @Override
-    public String getDisplayName()
-    {
-      return "Elastic Search";
+    public int hashCode() {
+        final int prime = 31;
+        int result = super.hashCode();
+        result = prime * result + ((uri == null) ? 0 : uri.hashCode());
+        result = prime * result + ((username == null) ? 0 : username.hashCode());
+        result = prime * result + Secret.toString(password).hashCode();
+        return result;
     }
 
     @Override
-    public int getDefaultPort()
-    {
-      return 0;
-    }
-
-    public FormValidation doCheckUrl(@QueryParameter("value") String value)
-    {
-      if (StringUtils.isBlank(value))
-      {
-        return FormValidation.warning(Messages.PleaseProvideHost());
-      }
-      try
-      {
-        URL url = new URL(value);
-
-        if (url.getUserInfo() != null)
-        {
-          return FormValidation.error("Please specify user and password not as part of the url.");
+    public ElasticSearchDao createIndexerInstance() {
+        ElasticSearchDao esDao = null;
+        HttpClientBuilder httpClientBuilder = HttpClientBuilder.create();
+
+        // Install custom certificate (if present) on a best-effort basis
+        if (!StringUtils.isBlank(customServerCertificateId)) {
+            StandardCertificateCredentials certificateCredentials = getCredentials(customServerCertificateId);
+            if (certificateCredentials instanceof StandardCertificateCredentials) {
+                try {
+                    // Fetch our custom certificate
+                    KeyStore keyStore = certificateCredentials.getKeyStore();
+                    String alias = keyStore.aliases().nextElement();
+                    X509Certificate certificate = (X509Certificate) keyStore.getCertificate(alias);
+
+                    SSLContext sslContext = ElasticSearch.createSSLContext(alias, certificate);
+                    if (sslContext != null)
+                        httpClientBuilder = httpClientBuilder.setSslcontext(sslContext);
+                } catch (KeyStoreException | KeyManagementException | NoSuchAlgorithmException
+                        | UnrecoverableKeyException | CertificateException | IOException ex) {
+                    //TODO: Report exception in logger
+                }
+            }
         }
 
-        if(StringUtils.isBlank(url.getPath()) || url.getPath().trim().matches("^\\/+$")) {
-          return FormValidation.warning("Elastic Search requires a key to be able to index the logs.");
+        esDao = new ElasticSearchDao(httpClientBuilder, getUri(), username, Secret.toString(password));
+
+        esDao.setMimeType(getMimeType());
+        return esDao;
+    }
+
+    public static SSLContext createSSLContext(String alias, X509Certificate certificate)
```

Comment:
I think creating the ssl context and the httpclient should not be part of the configuration class but part of the dao.

Reply:
@mwinter69 seems like a good idea - I've refactored the classes such that the SSL context creation and setting are now part of the DAO, while the credential access and interfacing with Jenkins are part of the config object.